clients/hadoopfs/src/main/java/io/lakefs/storage/LakeFSFileSystemOutputStream.java

```diff
@@ -0,0 +1,81 @@
+package io.lakefs.storage;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.hadoop.fs.Path;
+
+import io.lakefs.LakeFSClient;
+import io.lakefs.LakeFSFileSystem;
+import io.lakefs.clients.api.ApiException;
+import io.lakefs.clients.api.StagingApi;
+import io.lakefs.clients.api.model.StagingLocation;
+import io.lakefs.clients.api.model.StagingMetadata;
+import io.lakefs.utils.ObjectLocation;
+
+/**
+ * Handle writes into a storage URL.
+ * Will set the request Content-Length header.
+ * When closed, links the address in lakeFS.
+ * 
+ * TODO(johnnyaug): do not hold everything in memory
+ * TODO(johnnyaug): support multipart uploads
+ */
+public class LakeFSFileSystemOutputStream extends OutputStream {
+    private final HttpURLConnection connection;
+    private final ByteArrayOutputStream buffer;
+    private final StagingLocation stagingLocation;
+    private final ObjectLocation objectLoc;
+    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private final LakeFSFileSystem lfs;
+    private final LakeFSClient lakeFSClient;
+
+    public LakeFSFileSystemOutputStream(LakeFSFileSystem lfs, LakeFSClient lfsClient, HttpURLConnection connection, ObjectLocation objectLoc, StagingLocation stagingLocation) throws IOException {
+        this.connection = connection;
+        this.objectLoc = objectLoc;
+        this.stagingLocation = stagingLocation;
+        this.lfs = lfs;
+        this.lakeFSClient = lfsClient;
+        this.buffer = new ByteArrayOutputStream();
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        buffer.write(b);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (isClosed.getAndSet(true)) {
+            return;
+        }
+        connection.setRequestProperty("Content-Length", String.valueOf(buffer.size()));
```

Comment:
Nit: different capitalization here and on the next line.

Reply:
Agreed this is annoying, but still want to follow the [docs by Microsoft](https://learn.microsoft.com/en-us/rest/api/storageservices/put-blob?tabs=azure-ad#request-headers-all-blob-types) (even though headers _should_ be case insensitive).